// @bigcommerce/big-design/dist/es/components/Panel/Panel.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import React, { forwardRef, memo } from 'react';
import { excludePaddingProps } from '../../mixins/paddings/paddings';
import { Button } from '../Button';
import { Flex } from '../Flex';
import { StyledH2, StyledPanel } from './styled';
export var RawPanel = /*#__PURE__*/memo(function (_ref) {
  var forwardedRef = _ref.forwardedRef,
      props = _objectWithoutProperties(_ref, ["forwardedRef"]);

  var filteredProps = excludePaddingProps(props);

  var action = filteredProps.action,
      children = filteredProps.children,
      header = filteredProps.header,
      rest = _objectWithoutProperties(filteredProps, ["action", "children", "header"]);

  var renderHeader = function renderHeader() {
    if (!header && !action) {
      return null;
    }

    if (typeof header !== 'string') {
      return null;
    }

    return /*#__PURE__*/React.createElement(Flex, {
      flexDirection: "row"
    }, header && /*#__PURE__*/React.createElement(StyledH2, null, header), action && /*#__PURE__*/React.createElement(Button, action, action.text));
  };

  return /*#__PURE__*/React.createElement(StyledPanel, _extends({}, rest, {
    backgroundColor: "white",
    shadow: "raised",
    padding: {
      mobile: 'medium',
      tablet: 'xxLarge'
    },
    borderRadius: "none",
    ref: forwardedRef
  }), renderHeader(), children);
});
export var Panel = /*#__PURE__*/forwardRef(function (_ref2, ref) {
  var className = _ref2.className,
      style = _ref2.style,
      props = _objectWithoutProperties(_ref2, ["className", "style"]);

  return /*#__PURE__*/React.createElement(RawPanel, _extends({}, props, {
    forwardedRef: ref
  }));
});
Panel.displayName = 'Panel';
