// @bigcommerce/big-design/dist/cjs/components/Typography/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function get() {
    return _Typography.Text;
  }
});
Object.defineProperty(exports, "Small", {
  enumerable: true,
  get: function get() {
    return _Typography.Small;
  }
});
Object.defineProperty(exports, "HR", {
  enumerable: true,
  get: function get() {
    return _Typography.HR;
  }
});
Object.defineProperty(exports, "H0", {
  enumerable: true,
  get: function get() {
    return _Typography.H0;
  }
});
Object.defineProperty(exports, "H1", {
  enumerable: true,
  get: function get() {
    return _Typography.H1;
  }
});
Object.defineProperty(exports, "H2", {
  enumerable: true,
  get: function get() {
    return _Typography.H2;
  }
});
Object.defineProperty(exports, "H3", {
  enumerable: true,
  get: function get() {
    return _Typography.H3;
  }
});
Object.defineProperty(exports, "H4", {
  enumerable: true,
  get: function get() {
    return _Typography.H4;
  }
});

var _Typography = require("./Typography");
