// @bigcommerce/big-design/dist/es/components/Button/styled.js
import { addValues, theme as defaultTheme } from '@bigcommerce/big-design-theme';
import styled, { css } from 'styled-components';
import { withMargins } from '../../mixins';
import { withTransition } from '../../mixins/transitions';
import { Flex } from '../Flex';
export var StyledButton = /*#__PURE__*/styled.button.withConfig({
  displayName: "styled__StyledButton",
  componentId: "sc-3yq204-0"
})(["", " &&{", ";}align-items:center;appearance:none;border:", ";border-radius:", ";color:", ";cursor:pointer;display:inline-flex;flex:none;font-size:", ";font-weight:", ";height:", ";justify-content:center;line-height:", ";outline:none;padding:", ";pointer-events:", ";position:relative;text-align:center;text-decoration:none;user-select:none;vertical-align:middle;white-space:nowrap;width:", ";&:focus{outline:none;}&[disabled]{border-color:", ";pointer-events:none;}& + .bd-button{margin-top:", ";margin-left:", ";", "{margin-top:", ";margin-left:", ";}}", "{width:auto;}", ";", ";", ";", ""], withTransition(['background-color', 'border-color', 'box-shadow', 'color']), withMargins(), function (_ref) {
  var theme = _ref.theme;
  return theme.border.box;
}, function (_ref2) {
  var theme = _ref2.theme;
  return theme.borderRadius.normal;
}, function (_ref3) {
  var theme = _ref3.theme;
  return theme.colors.white;
}, function (_ref4) {
  var theme = _ref4.theme;
  return theme.typography.fontSize.medium;
}, function (_ref5) {
  var theme = _ref5.theme;
  return theme.typography.fontWeight.regular;
}, function (_ref6) {
  var theme = _ref6.theme;
  return addValues(theme.spacing.xxLarge, theme.spacing.xxSmall);
}, function (_ref7) {
  var theme = _ref7.theme;
  return theme.lineHeight.xLarge;
}, function (_ref8) {
  var theme = _ref8.theme;
  return "0 ".concat(theme.spacing.medium);
}, function (_ref9) {
  var isLoading = _ref9.isLoading;
  return isLoading ? 'none' : 'auto';
}, function (_ref10) {
  var mobileWidth = _ref10.mobileWidth;
  return mobileWidth === 'auto' ? 'auto' : '100%';
}, function (_ref11) {
  var theme = _ref11.theme;
  return theme.colors.secondary30;
}, function (_ref12) {
  var mobileWidth = _ref12.mobileWidth,
      theme = _ref12.theme;
  return mobileWidth === '100%' && theme.spacing.xSmall;
}, function (_ref13) {
  var mobileWidth = _ref13.mobileWidth,
      theme = _ref13.theme;
  return mobileWidth === 'auto' && theme.spacing.xSmall;
}, function (_ref14) {
  var theme = _ref14.theme;
  return theme.breakpoints.tablet;
}, function (_ref15) {
  var theme = _ref15.theme;
  return theme.spacing.none;
}, function (_ref16) {
  var theme = _ref16.theme;
  return theme.spacing.xSmall;
}, function (_ref17) {
  var theme = _ref17.theme;
  return theme.breakpoints.tablet;
}, function (_ref18) {
  var icon = _ref18.iconOnly,
      theme = _ref18.theme;
  return icon && css(["padding:0 ", ";"], theme.spacing.xSmall);
}, function (_ref19) {
  var iconLeft = _ref19.iconLeft,
      theme = _ref19.theme;
  return iconLeft && css(["padding-left:", ";"], theme.spacing.xSmall);
}, function (_ref20) {
  var iconRight = _ref20.iconRight,
      theme = _ref20.theme;
  return iconRight && css(["padding-right:", ";"], theme.spacing.xSmall);
}, function (props) {
  return getButtonStyles(props);
});
export var ContentWrapper = /*#__PURE__*/styled.span.attrs({}).withConfig({
  displayName: "styled__ContentWrapper",
  componentId: "sc-3yq204-1"
})(["align-content:center;align-items:center;display:inline-grid;grid-auto-flow:column;grid-gap:", ";visibility:", ";"], function (_ref21) {
  var theme = _ref21.theme;
  return theme.spacing.xSmall;
}, function (_ref22) {
  var isLoading = _ref22.isLoading;
  return isLoading ? 'hidden' : 'visible';
});
export var LoadingSpinnerWrapper = /*#__PURE__*/styled(Flex).withConfig({
  displayName: "styled__LoadingSpinnerWrapper",
  componentId: "sc-3yq204-2"
})(["position:absolute;"]);
/**
 * These can be generated dynamically but I'm leaning towards being extra
 * explicit and being able to handle corner cases and changes from design easily
 */

var ButtonPrimary = /*#__PURE__*/css(["background-color:", ";border-color:", ";font-weight:", ";&:active{background-color:", ";}&:focus{box-shadow:", ";}&:hover:not(:active){background-color:", ";}&[disabled]{background-color:", ";}"], function (_ref23) {
  var theme = _ref23.theme;
  return theme.colors.primary;
}, function (_ref24) {
  var theme = _ref24.theme;
  return theme.colors.primary;
}, function (_ref25) {
  var theme = _ref25.theme;
  return theme.typography.fontWeight.semiBold;
}, function (_ref26) {
  var theme = _ref26.theme;
  return theme.colors.primary60;
}, function (_ref27) {
  var theme = _ref27.theme;
  return "0 0 0 ".concat(theme.spacing.xxSmall, " ").concat(theme.colors.primary20);
}, function (_ref28) {
  var theme = _ref28.theme;
  return theme.colors.primary50;
}, function (_ref29) {
  var theme = _ref29.theme;
  return theme.colors.secondary30;
});
var ButtonPrimaryDestructive = /*#__PURE__*/css(["background-color:", ";border-color:", ";font-weight:", ";&:active{background-color:", ";}&:focus{box-shadow:", ";}&:hover:not(:active){background-color:", ";}&[disabled]{background-color:", ";}"], function (_ref30) {
  var theme = _ref30.theme;
  return theme.colors.danger;
}, function (_ref31) {
  var theme = _ref31.theme;
  return theme.colors.danger;
}, function (_ref32) {
  var theme = _ref32.theme;
  return theme.typography.fontWeight.semiBold;
}, function (_ref33) {
  var theme = _ref33.theme;
  return theme.colors.danger60;
}, function (_ref34) {
  var theme = _ref34.theme;
  return "0 0 0 ".concat(theme.spacing.xxSmall, " ").concat(theme.colors.danger20);
}, function (_ref35) {
  var theme = _ref35.theme;
  return theme.colors.danger50;
}, function (_ref36) {
  var theme = _ref36.theme;
  return theme.colors.secondary30;
});
var ButtonSecondary = /*#__PURE__*/css(["background-color:transparent;border-color:", ";color:", ";&:active{background-color:", ";}&:focus{box-shadow:", ";}&:hover:not(:active){background-color:", ";}&[disabled]{color:", ";}"], function (_ref37) {
  var theme = _ref37.theme;
  return theme.colors.primary;
}, function (_ref38) {
  var theme = _ref38.theme;
  return theme.colors.primary;
}, function (_ref39) {
  var theme = _ref39.theme;
  return theme.colors.primary20;
}, function (_ref40) {
  var theme = _ref40.theme;
  return "0 0 0 ".concat(theme.spacing.xxSmall, " ").concat(theme.colors.primary20);
}, function (_ref41) {
  var theme = _ref41.theme;
  return theme.colors.primary10;
}, function (_ref42) {
  var theme = _ref42.theme;
  return theme.colors.secondary30;
});
var ButtonSecondaryDestructive = /*#__PURE__*/css(["background-color:transparent;border-color:", ";color:", ";&:active{background-color:", ";}&:focus{box-shadow:", ";}&:hover:not(:active){background-color:", ";}&[disabled]{color:", ";}"], function (_ref43) {
  var theme = _ref43.theme;
  return theme.colors.danger;
}, function (_ref44) {
  var theme = _ref44.theme;
  return theme.colors.danger;
}, function (_ref45) {
  var theme = _ref45.theme;
  return theme.colors.danger20;
}, function (_ref46) {
  var theme = _ref46.theme;
  return "0 0 0 ".concat(theme.spacing.xxSmall, " ").concat(theme.colors.danger20);
}, function (_ref47) {
  var theme = _ref47.theme;
  return theme.colors.danger10;
}, function (_ref48) {
  var theme = _ref48.theme;
  return theme.colors.secondary30;
});
var ButtonSubtle = /*#__PURE__*/css(["background-color:transparent;border-color:transparent;color:", ";&:active{background-color:", ";}&:focus{box-shadow:", ";}&:hover:not(:active){background-color:", ";}&[disabled]{border-color:transparent;color:", ";}"], function (_ref49) {
  var theme = _ref49.theme;
  return theme.colors.primary;
}, function (_ref50) {
  var theme = _ref50.theme;
  return theme.colors.primary20;
}, function (_ref51) {
  var theme = _ref51.theme;
  return "0 0 0 ".concat(theme.spacing.xxSmall, " ").concat(theme.colors.primary20);
}, function (_ref52) {
  var theme = _ref52.theme;
  return theme.colors.primary10;
}, function (_ref53) {
  var theme = _ref53.theme;
  return theme.colors.secondary30;
});
var ButtonSubtleDestructive = /*#__PURE__*/css(["background-color:transparent;border-color:transparent;color:", ";&:active{background-color:", ";}&:focus{box-shadow:", ";}&:hover:not(:active){background-color:", ";}&[disabled]{border-color:transparent;color:", ";}"], function (_ref54) {
  var theme = _ref54.theme;
  return theme.colors.danger;
}, function (_ref55) {
  var theme = _ref55.theme;
  return theme.colors.danger20;
}, function (_ref56) {
  var theme = _ref56.theme;
  return "0 0 0 ".concat(theme.spacing.xxSmall, " ").concat(theme.colors.danger20);
}, function (_ref57) {
  var theme = _ref57.theme;
  return theme.colors.danger10;
}, function (_ref58) {
  var theme = _ref58.theme;
  return theme.colors.secondary30;
});

function getButtonStyles(props) {
  var actionType = props.actionType,
      variant = props.variant;

  switch (variant) {
    case 'primary':
      return actionType === 'destructive' ? ButtonPrimaryDestructive : ButtonPrimary;

    case 'secondary':
      return actionType === 'destructive' ? ButtonSecondaryDestructive : ButtonSecondary;

    case 'subtle':
      return actionType === 'destructive' ? ButtonSubtleDestructive : ButtonSubtle;
  }
}

StyledButton.defaultProps = {
  theme: defaultTheme
};
ContentWrapper.defaultProps = {
  theme: defaultTheme
};
LoadingSpinnerWrapper.defaultProps = {
  theme: defaultTheme
};
